sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/JavaBeanSchemaTest.java

```diff
@@ -155,10 +155,10 @@ public void testRecursiveGetters() throws NoSuchSchemaException {
 
     Row nestedRow = row.getRow("nested");
     assertEquals("string", nestedRow.getString("str"));
-    assertEquals((byte) 1, nestedRow.getByte("aByte"));
-    assertEquals((short) 2, nestedRow.getInt16("aShort"));
-    assertEquals((int) 3, nestedRow.getInt32("anInt"));
-    assertEquals((long) 4, nestedRow.getInt64("aLong"));
+    assertEquals((byte) 1, (Object) nestedRow.getByte("aByte"));
```

Comment:
Why not do `(Byte) 1, (Byte) nestedRow.getByte("aByte")`?

Reply:
1. Because you can't cast a constant directly to `(Byte)` in java 8.
2. Because I want to make sure getByte is returning the `Byte` type as part of this test and not just something that can promote to `Byte`.

Reply:
1. I think that `Byte.valueOf(byte)` is meant for this;
2. I may be wrong, but I think that the only thing that can promote to `Byte` is `byte`, I don't believe you can cast boxed types, like `(Double)(Integer) getBlah()`;

Additionally, it looks like the first parameter will be boxed by `assertEquals(Object, Object)` anyway, so essentially with a cast (either `(Byte)` or `(Object)`, doesn't matter) or without a cast this should be equivalent to `assertTrue(Byte.valueOf(1).equals(nestedRow.getByte("aByte")))`, I may be missing something though.

Plus this test is not ensuring that it's actually a `Byte`, not `byte`. I would probably add a test that returns `null`.

Reply:
Undoing this diff results in the following failure:
```
JavaBeanSchemaTest.java:158: error: reference to assertEquals is ambiguous
    assertEquals((byte) 1, nestedRow.getByte("aByte"));
    ^
  both method assertEquals(long,long) in Assert and method assertEquals(Object,Object) in Assert match
```

@amaliuja Your suggestion results in the following failure:
```
JavaBeanSchemaTest.java:158: error: incompatible types: int cannot be converted to Byte
    assertEquals((Byte) 1, (Byte) nestedRow.getByte("aByte"));
```

@akedin Your suggestion results in the following failure:
```
    assertEquals(Byte.valueOf(1), nestedRow.getByte("aByte"));
                     ^
    method Byte.valueOf(byte) is not applicable
      (argument mismatch; possible lossy conversion from int to byte)
    method Byte.valueOf(String) is not applicable
      (argument mismatch; int cannot be converted to String)
```
Using `assertTrue` is incorrect because it does not print the values on failure. I'm happy to change the syntax, but none of these options work.

Reply:
It is unfortunate that there's `assertEquals(long, long)` overload that is applicable here.

Correct way of doing `Byte.valueOf()` is `Byte.valueOf((byte) 1)`, which is ugly. Probably even uglier way is `(Byte)(byte)1`. Another ugly approach is to create a local wrapper method `void myAssertEquals(Object a, Object b) { assertEquals(a, b); }`.

I mentioned `assertTrue()` as an example, not suggesting to use it.

To Rui's point, though, if you look at how `Byte.equals()` works, it accepts an `Object` and casts it to `Byte`, so in this case it shouldn't matter where to cast it at all, to cast it to `Byte` or `Object`. Casting here only allows to unambiguously choose the `assertEquals(Object, Object)` overload over `assertEquals(long, long)`, and I don't have an idea how to do that cleanly, so I don't mind either your approach or any of the proposed in the discussion.

Reply:
or Byte.valueof("1") ?

Reply:
My intention was to make the code more understandable for readers. Because if we do explicitly casting and leave less guess room for readers, readers will know the same type of values are compared, and they do not need to question the correctness because of implicitly casting somewhere. At least, readers do not need to search  how `assertEquals` is implemented for different types. From this perspective, `assertEquals(Byte.valueof((byte) 1), nestedRow.getByte("aByte"))` is ugly, but necessary.

After I read the code, seems like the family of `assertEquals` in JUnit is simpler. `assertEquals(Byte.valueof((byte) 1), nestedRow.getByte("aByte"))` will be converted `assertEquals(object, object)`, and `object.equals(object)` will be called. In `Byte` implementation, as Anton pointed out, `Byte.equals()` accepts `Object` and does a casting anyway. So from implementation perspective, there is no difference among approaches in this discussion.

Since there isn't a perfect way to make the code clear, so I am ok with either the original way in this PR, or other discussed way to implement it.